File: lib/collapse/collapse.dart
import "package:angular2/angular2.dart";
import 'dart:async';
// todo: add animate

// todo: add init and on change
@Directive (selector: "[collapse]",
    inputs: const [ "collapse"],
    host: const {
      "[class.in]" : "isExpanded",
      "[class.collapse]" : "isCollapse",
      "[class.collapsing]" : "isCollapsing",
      "[attr.aria-expanded]" : "isExpanded",
      "[attr.aria-hidden]" : "isCollapsed",
      "[style.height]" : "height"
    })
class Collapse {
  ElementRef el;

  dynamic test = "wtf";

  // style
  String height;

  // classes

  // shown
  bool isExpanded = true;

  // hidden
  bool isCollapsed = false;

  // stale state
  bool isCollapse = true;

  // animation state
  bool isCollapsing = false;

  Collapse(this.el);

  bool get collapse {
    return isExpanded;
  }

  set collapse(bool value) {
    isExpanded = value;
    toggle();
  }

  toggle() {
    if (isExpanded) {
      hide();
    } else {
      show();
    }
  }

  hide() {
    isCollapse = false;
    isCollapsing = true;
    isExpanded = false;
    isCollapsed = true;
    new Timer(const Duration(milliseconds: 4), () {
      height = "0";
      isCollapse = true;
      isCollapsing = false;
    });
  }

  show() {
    isCollapse = false;
    isCollapsing = true;
    isExpanded = true;
    isCollapsed = false;
    new Timer(const Duration(milliseconds: 4), () {
      height = "auto";
      isCollapse = true;
      isCollapsing = false;
    });
  }
}

File: components/collapse/collapse.ts
import {
  Directive, Input, HostBinding
} from 'angular2/core';

// todo: add animate
// todo: add init and on change
@Directive({ selector: '[collapse]' })
export class Collapse {
  // style
  @HostBinding('style.height')
  private height:string;
  // shown
  @HostBinding('class.in')
  @HostBinding('attr.aria-expanded')
  private isExpanded:boolean = true;
  // hidden
  @HostBinding('attr.aria-hidden')
  private isCollapsed:boolean = false;
  // stale state
  @HostBinding('class.collapse')
  private isCollapse:boolean = true;
  // animation state
  @HostBinding('class.collapsing')
  private isCollapsing:boolean = false;

  @Input()
  private set collapse(value:boolean) {
    this.isExpanded = value;
    this.toggle();
  }

  private get collapse():boolean {
    return this.isExpanded;
  }

  constructor() {
  }

  toggle() {
    if (this.isExpanded) {
      this.hide();
    } else {
      this.show();
    }
  }

  hide() {
    this.isCollapse = false;
    this.isCollapsing = true;

    this.isExpanded = false;
    this.isCollapsed = true;
    setTimeout(() => {
      this.height = '0';
      this.isCollapse = true;
      this.isCollapsing = false;
    }, 4);
  }

  show() {
    this.isCollapse = false;
    this.isCollapsing = true;

    this.isExpanded = true;
    this.isCollapsed = false;
    setTimeout(() => {
      this.height = 'auto';

      this.isCollapse = true;
      this.isCollapsing = false;
    }, 4);
  }
}
